### DISCUSSION TITLE
Firedrake installation on Ubuntu 20.04 - cmake/libsupermesh issues

### DISCUSSION THREAD

**Discussion by cpjordan** [firedrakeproject/firedrake#3374] [Firedrake support] [answered]
I've tried to install firedrake on a new workstation (Ubuntu 20.04 OS) but have run into problems. I attach the relevant logs:
[firedrake-install.log](https://github.com/firedrakeproject/firedrake/files/14100529/firedrake-install.log)
[configure.log](https://github.com/firedrakeproject/firedrake/files/14100583/configure.log)
[make.log](https://github.com/firedrakeproject/firedrake/files/14100588/make.log)

The issue initially seemed to be with cmake so I started from scratch and pre-installed the latest version (3.28) before installing Firedrake again, but it seemed to ignore it when building Firedrake. Instead I just updated cmake after installing Firedrake and tried to do firedrake-update instead, which seemed to solve that problem (I think). Now, however, I have an issue with not finding libsupermesh-c.h, which is definitely present in firedrake/src/libsupermesh/include. firedrake-update.log attached:
[firedrake-update.log](https://github.com/firedrakeproject/firedrake/files/14100695/firedrake-update.log)

Any help would be massively appreciated! I am using Python 3.8, I am not using Anaconda and I am using the latest firedrake install script.

**Comment by connorjward** [comment]
@JDBetteridge you tinkered with the cmake bits of libsupermesh recently. Do you have any insights? It appears to be picking up the wrong cmake executable as earlier in the script cmake 3.27.9 is installed by PETSc.

**Comment by JDBetteridge** [comment]
The CMake that is installed by PETSc isn't on the `PATH`. We should probably detect this and set the CMake installed by PETSc to be the CMake we use in the rest of the install script.

A quick fix would be to activate the Firedrake virtual environment, install libsupermesh manually and then update, by running:
```
cd $VIRTUAL_ENV/src/libsupermesh/build
$VIRTUAL_ENV/src/petsc/default/bin/cmake .. \
    -DBUILD_SHARED_LIBS=ON \
    -DCMAKE_INSTALL_PREFIX=$VIRTUAL_ENV \
    -DMPI_C_COMPILER=$VIRTUAL_ENV/bin/mpicc \
    -DMPI_CXX_COMPILER=$VIRTUAL_ENV/bin/mpicxx \
    -DMPI_Fortran_COMPILER=$VIRTUAL_ENV/bin/mpif90 \
    -DCMAKE_Fortran_COMPILER=$VIRTUAL_ENV/bin/mpif90 \
    -DMPIEXEC_EXECUTABLE=$VIRTUAL_ENV/bin/mpiexec
make
make install
firedrake-update
```

Double check this, I might have made a typo

**Reply by cpjordan** [reply]
Brilliant, that's done the job - thank you so much for the help!

**Comment by cpjordan** [comment]
I tried a fresh install of Firedrake this morning and bumped into the same problem, except this time the above solution no longer seems to work. I've attached the relevant files. Python 3.11 & Ubuntu 20.04.

[firedrake-install.log](https://github.com/user-attachments/files/18299849/firedrake-install.log)
[firedrake/firedrake-update.log](https://github.com/user-attachments/files/18299854/firedrake-update.log)
[src/petsc/configure.log](https://github.com/user-attachments/files/18299858/configure.log)
[src/petsc/make.log](https://github.com/user-attachments/files/18299862/make.log)

Could you please let me know how to link the `libsupermesh` install to the correct `cmake` in `$VIRTUAL_ENV/src/petsc/default/bin/cmake` in the Firedrake install/update? I can't upgrade the machine I am using above Ubuntu 20.04 due to some frustrating system memory issues, so it would also be nice to link to the correct `cmake` version in the install if possible so that users with older Ubuntu versions can install without this, if possible!

**Reply by connorjward** [reply, marked as answer]
I think you simply need to run
```
$ export PATH=$VIRTUAL_ENV/src/petsc/default/bin:$PATH
$ firedrake-update
```

`firedrake-install` will be deprecated within the next month or so, at which point this should stop being an issue.
